src/Common/FiberStack.cpp

```diff
@@ -0,0 +1,242 @@
+#if defined(BOOST_USE_VALGRIND)
+    sctx.valgrind_stack_id = VALGRIND_STACK_REGISTER(sctx.sp, vp);
+#endif
+    return sctx;
+}
+
+void FiberStack::deallocate(boost::context::stack_context & sctx) const
+{
+    ProfileEvents::increment(ProfileEvents::FiberStackDeallocationCount);
+    Stopwatch watch;
+    SCOPE_EXIT(
+        ProfileEvents::increment(ProfileEvents::FiberStackDeallocationMicroseconds, watch.elapsedMicroseconds());
+    );
+
+#if defined(BOOST_USE_VALGRIND)
+    VALGRIND_STACK_DEREGISTER(sctx.valgrind_stack_id);
+#endif
+    void * vp = static_cast< char * >(sctx.sp) - sctx.size;
+    ::munmap(vp, sctx.size);
+
+    /// Do not count guard page in memory usage.
+    auto trace = CurrentMemoryTracker::free(sctx.size - page_size);
+    trace.onFree(vp, sctx.size - page_size);
+}
+
+/// A pool of fixed size fiber stacks, thread safe.
+/// Other threads can borrow stacks from this pool and return them once done.
+class FixedSizeFiberStackCache
+{
+private:
+    std::mutex mutex;
+    size_t max_cached_bytes; /// Maximum number of cached bytes.
+    FiberStack * allocator TSA_GUARDED_BY(mutex); /// Base allocator for new stacks.
+    std::list<boost::context::stack_context> free_stacks TSA_GUARDED_BY(mutex); /// List of free stacks.
+    absl::flat_hash_set<void *> allocated_stacks TSA_GUARDED_BY(mutex); /// Map of allocated pointer -> allocated_stack.
+    size_t cached_bytes TSA_GUARDED_BY(mutex);
+#ifndef NDEBUG
+    size_t page_size;
+#endif
+
+public:
+    explicit FixedSizeFiberStackCache(FiberStack * base_allocator_, size_t max_cached_bytes_)
+        : max_cached_bytes(max_cached_bytes_), allocator(base_allocator_), cached_bytes(0)
+    {
+#ifndef NDEBUG
+        page_size = getPageSize();
+#endif
+    }
+    /// Borrow a stack from the pool.
+    /// If no free stacks are available, allocate a new one.
+    /// If the maximum number of cached stacks is reached, return nullopt.
+    std::optional<boost::context::stack_context> borrowStack()
+    {
+        std::lock_guard lock(mutex);
+        if (free_stacks.empty())
+        {
+            if (cached_bytes + FiberStack::default_stack_size >= max_cached_bytes)
+                return std::nullopt;
+            free_stacks.push_front(allocator->allocate());
+            CurrentMetrics::add(CurrentMetrics::FiberStackCacheBytes, free_stacks.front().size);
+        }
+        auto stack = free_stacks.front();
+        free_stacks.pop_front();
+        allocated_stacks.insert(stack.sp);
+
+    /// TODO: do we need to zero out the stack in all cases since all fiber is in same process?
+    /// Folly::GuardPageAllocator, boost::context::pooled_fixedsize_stack, boost::coroutines::standard_stack_allocator
+    /// don't zero out the stack. glibc pthread zero out the thread local storage section of the stack.
+#ifndef NDEBUG
+        auto * vp = static_cast<char *>(stack.sp) - stack.size;
+        ::memset(vp + page_size, 0, stack.size - page_size);
```

Comment:
we are still in critical section here. I'm sure that we could `memset` outside of critical section.